crates/redeem/src/lib.rs

```diff
@@ -501,23 +502,48 @@ impl<T: Config> Pallet<T> {
             Error::<T>::AmountBelowDustAmount
         );
 
-        // vault will get rid of the btc + btc_inclusion_fee
-        ext::vault_registry::try_increase_to_be_redeemed_tokens::<T>(&vault_id, &vault_to_be_burned_tokens)?;
-
-        // lock full amount (inc. fee)
-        amount_wrapped.lock_on(&redeemer)?;
-        let redeem_id = ext::security::get_secure_id::<T>(&redeemer);
-
         let below_premium_redeem = ext::vault_registry::is_vault_below_premium_threshold::<T>(&vault_id)?;
         let currency_id = vault_id.collateral_currency();
 
         let premium_collateral = if below_premium_redeem {
-            let redeem_amount_wrapped_in_collateral = user_to_be_received_btc.convert_to(currency_id)?;
-            ext::fee::get_premium_redeem_fee::<T>(&redeem_amount_wrapped_in_collateral)?
+            // Calculate the secure vault capacity
+            let secure_vault_capacity = ext::vault_registry::vault_capacity_at_secure_threshold(&vault_id)?;
+            let to_be_backed_tokens = ext::vault_registry::vault_to_be_backed_tokens(&vault_id)?;
+            let difference_in_tokens_to_reach_secure_threshold =
+                to_be_backed_tokens.saturating_sub(&secure_vault_capacity)?;
+
+            if difference_in_tokens_to_reach_secure_threshold.gt(&user_to_be_received_btc)? {
```

Comment:
Why do we need this `if`? Shouldn't the `else` branch always work?

Reply:
the formula gives maximum premium, the user shouldn't receive `maximum_premium`, if he has not redeem enough BTC. 
Example, test case: `integration_test_premium_redeem_with_reward_for_full_request`
alice requests 100 BTC redeem, vault can do 150 BTC redeem to reach secure threshold, therefore alice should only receive `premium_collateral` for 100BTC - tx fee ie  98.300 according to testcase, if we remove the if condition he will receive the max `premium_collateral`.